greenpithumb/greenpithumb.py

```diff
@@ -29,35 +29,135 @@
+
+
+def make_adc(wiring_config):
+    """Creates ADC instance based on the given wiring_config."""
     # The MCP3008 spec and Adafruit library use different naming for the
     # Raspberry Pi GPIO pins, so we translate as follows:
     # * CLK -> CLK
     # * CS/SHDN -> CS
     # * DOUT -> MISO
     # * DIN -> MOSI
-    adc = adc_thread_safe.Adc(
+    return adc_thread_safe.Adc(
         Adafruit_MCP3008.MCP3008(
             clk=wiring_config.gpio_pins.mcp3008_clk,
             cs=wiring_config.gpio_pins.mcp3008_cs_shdn,
             miso=wiring_config.gpio_pins.mcp3008_dout,
             mosi=wiring_config.gpio_pins.mcp3008_din))
+
+
+def make_dht11_sensors(wiring_config):
+    """Creates sensors derived from the DHT11 sensor.
+
+    Args:
+        wiring_config: Wiring configuration for the GreenPiThumb.
+
+    Returns:
+        A two-tuple where the first element is a temperature sensor and the
+        second element is a humidity sensor.
+    """
     local_dht11 = dht11.CachingDHT11(
         lambda: Adafruit_DHT.read_retry(Adafruit_DHT.DHT11, wiring_config.gpio_pins.dht11),
-        utc_clock)
-    water_pump = pump.Pump(raspberry_pi_io, utc_clock,
-                           wiring_config.gpio_pins.pump)
-    pump_scheduler = pump.PumpScheduler(local_clock, sleep_windows)
-    pump_timer = clock.Timer(utc_clock, pump_interval)
-    pump_manager = pump.PumpManager(water_pump, pump_scheduler,
-                                    moisture_threshold, pump_amount, pump_timer)
+        clock.Clock())
+    return temperature_sensor.TemperatureSensor(
+        local_dht11), humidity_sensor.HumiditySensor(local_dht11),
+
+
+def make_soil_moisture_sensor(adc, raspberry_pi_io, wiring_config):
+    return soil_moisture_sensor.SoilMoistureSensor(
+        adc, raspberry_pi_io, wiring_config.adc_channels.soil_moisture_sensor,
+        wiring_config.gpio_pins.soil_moisture)
+
+
+def make_light_sensor(adc, wiring_config):
+    return light_sensor.LightSensor(adc,
+                                    wiring_config.adc_channels.light_sensor)
+
+
+def make_camera_manager(rotation, image_path, light_sensor):
+    """Creates a camera manager instance.
+
+    Args:
+        rotation: The amount (in whole degrees) to rotate the camera image.
+        image_path: The directory in which to save images.
+        light_sensor: A light sensor instance.
+
+    Returns:
+        A CameraManager instance with the given settings.
+    """
+    camera = picamera.PiCamera(resolution=picamera.PiCamera.MAX_RESOLUTION)
+    camera.rotation = rotation
+    return camera_manager.CameraManager(image_path,
+                                        clock.Clock(), camera, light_sensor)
+
+
+def make_pump_manager(moisture_threshold, sleep_windows, raspberry_pi_io,
+                      wiring_config, pump_amount, pump_interval):
+    """Creates a pump manager instance.
+
+    Args:
+        moisture_threshold: The minimum water level before the pump turns on.
```

Comment:
The way this is worded is a little confusing. Maybe something like 'The minimum moisture level below which the pump turns on.'